rl4co/envs/scheduling/fjsp/env.py

```diff
@@ -0,0 +1,436 @@
+        self.generator = generator
+        self.num_mas = generator.num_mas
+        self.num_jobs = generator.num_jobs
+        self.n_ops_max = generator.max_ops_per_job * self.num_jobs
+        self.mask_no_ops = mask_no_ops
+        self._make_spec(self.generator)
+
+    def _set_seed(self, seed):
+        """Set the seed for the environment"""
+        rng = torch.manual_seed(seed)
+        self.rng = rng
+
+    def _decode_graph_structure(self, td: TensorDict):
+        batch_size = td.batch_size
+        start_op_per_job = td["start_op_per_job"]
+        end_op_per_job = td["end_op_per_job"]
+        pad_mask = td["pad_mask"]
+        n_ops_max = td["pad_mask"].size(-1)
+
+        # here we will generate the operations-job mapping:
+        ops_job_map, ops_job_bin_map = get_job_ops_mapping(
+            start_op_per_job, end_op_per_job, n_ops_max
+        )
+
+        # mask invalid edges (caused by padding)
+        ops_job_bin_map[pad_mask.unsqueeze(1).expand_as(ops_job_bin_map)] = 0
+
+        # generate for each batch a sequence specifying the position of all operations in their respective jobs,
+        # e.g. [0,1,0,0,1,2,0,1,2,3,0,0] for jops with n_ops=[2,1,3,4,1,1]
+        # (bs, max_ops)
+        ops_seq_order = torch.sum(
+            ops_job_bin_map * (ops_job_bin_map.cumsum(2) - 1), dim=1
+        )
+
+        # predecessor and successor adjacency matrices
+        pred = torch.diag_embed(torch.ones(n_ops_max - 1), offset=-1)[None].expand(
+            *batch_size, -1, -1
+        )
+        # the start of the sequence (of each job) does not have a predecessor, therefore we can
+        # mask all first ops of a job in the predecessor matrix
+        pred = pred * ops_seq_order.gt(0).unsqueeze(-1).expand_as(pred).to(pred)
+        succ = torch.diag_embed(torch.ones(n_ops_max - 1), offset=1)[None].expand(
+            *batch_size, -1, -1
+        )
+        # apply the same logic as above to mask the last op of a job, which does not have a successor. The last job of a job
+        # always comes before the 1st op of the next job, therefore performing a left shift of the ops seq tensor here
+        succ = succ * torch.cat(
+            (ops_seq_order[:, 1:], ops_seq_order.new_full((*batch_size, 1), 0)), dim=1
+        ).gt(0).to(succ).unsqueeze(-1).expand_as(succ)
+
+        # adjacency matrix = predecessors, successors and self loops
+        # (bs, max_ops, max_ops, 2)
+        ops_adj = torch.stack((pred, succ), dim=3)
+
+        td = td.update(
+            {
+                "ops_adj": ops_adj,
+                "job_ops_adj": ops_job_bin_map,
+                "ops_job_map": ops_job_map,
+                # "op_spatial_enc": ops_spatial_enc,
+                "ops_sequence_order": ops_seq_order,
+            }
+        )
+
+        return td, n_ops_max
+
+    def _reset(self, td: TensorDict = None, batch_size=None) -> TensorDict:
+        device = td.device
+        self.to(device)
```

Comment:
[Minor] This part is included in the base class~ https://github.com/ai4co/rl4co/blob/684dbdcb6af83936a7733e814f1c9a42a280cab8/rl4co/envs/common/base.py#L135-L143